### PR TITLE
Use CMake version in Mac strings file

### LINKED ISSUE #830: Wrong version in Mac nightly Info.plist
This was noticed in the nightlies, but is probably an issue with any Mac build from master currently.  The InfoPlist.strings file placed deep in the .app, in Contents\Resources\English.lproj\, still claims to be version 3.7.3.

### PR DESCRIPTION
**Pull request by asarium**
Fixes #830.

As always, I can't test this directly so I'll need someone with a Mac to check if this works.

### CODE REVIEW

**Comment by chief1983** [comment]
Contents/Info.plist after checking out this PR:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>BuildMachineOSBuild</key>
	<string>16A323</string>
	<key>CFBundleDevelopmentRegion</key>
	<string>English</string>
	<key>CFBundleIconFile</key>
	<string>FS2_Open</string>
	<key>CFBundleInfoDictionaryVersion</key>
	<string>6.0</string>
	<key>CFBundleName</key>
	<string>FS2_Open</string>
	<key>CFBundlePackageType</key>
	<string>APPL</string>
	<key>CFBundleSignature</key>
	<string>????</string>
	<key>CFBundleSupportedPlatforms</key>
	<array>
		<string>MacOSX</string>
	</array>
	<key>CFBundleVersion</key>
	<string></string>
	<key>DTCompiler</key>
	<string>com.apple.compilers.llvm.clang.1_0</string>
	<key>DTPlatformBuild</key>
	<string>8A218a</string>
	<key>DTPlatformVersion</key>
	<string>GM</string>
	<key>DTSDKBuild</key>
	<string>16A300</string>
	<key>DTSDKName</key>
	<string>macosx10.12</string>
	<key>DTXcode</key>
	<string>0800</string>
	<key>DTXcodeBuild</key>
	<string>8A218a</string>
	<key>LSApplicationCategoryType</key>
	<string>public.app-category.games</string>
	<key>LSMinimumSystemVersion</key>
	<string></string>
	<key>NSMainNibFile</key>
	<string>MainMenu</string>
	<key>NSPrincipalClass</key>
	<string>NSApplication</string>
	<key>SDL_FILESYSTEM_BASE_DIR_TYPE</key>
	<string>parent</string>
</dict>
</plist>
```

Doesn't look like it worked.

**Comment by asarium** [comment]
Uhh, that's the wrong file. The changes only configures the InfoPlist.strings file.

**Comment by chief1983** [comment]
Ah, well.  Nothing I have can even read that file now.  Just 11 bytes of binary gibberish to Sublime, nano, and cat.

**Comment by asarium** [comment]
Well, I have no idea how that would happen. The original file has some weird binary stuff at the front but I just copied that from the existing files. Maybe CMake can't handle that for some reason.

**Comment by chief1983** [comment]
That could be  the case.  I did have to handle some files very specifically in the nightly build Perl code because of their unique encodings.  One of these might be a file that was UTF-16.

Also, the original issue should have applied to both of the files, Info.plist and InfoPlist.strings.  Info.Plist in Contents has a missing version string, compared to the 3.7.4 release.  InfoPlist.strings has the old version string 3.7.3 instead of having been bumped to 3.7.5 at some point.  It also seems to be missing this:

```
	<key>CFBundleExecutable</key>
	<string>FS2_Open 3.7.4 (debug)</string>
```

Again, just compared to 3.7.4.

**Comment by asarium** [comment]
Is it possible to just use UTF-8 instead of UTF-16 BE?

Again, I can't find any version information in `Info.plist`, even when I am looking at the 3.7.4 tag. All I can find is
    
	<key>CFBundleExecutable</key>
	<string>${PRODUCT_NAME}</string>

when I check out the file from the release tag.

**Comment by chief1983** [comment]
I'm just going off of the completed bundles.  Here's the complete 3.7.4 Info.plist that ends up in a bundle:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>BuildMachineOSBuild</key>
	<string>15F34</string>
	<key>CFBundleDevelopmentRegion</key>
	<string>English</string>
	<key>CFBundleExecutable</key>
	<string>FS2_Open 3.7.4 (debug)</string>
	<key>CFBundleIconFile</key>
	<string>FS2_Open</string>
	<key>CFBundleInfoDictionaryVersion</key>
	<string>6.0</string>
	<key>CFBundleName</key>
	<string>FS2_Open</string>
	<key>CFBundlePackageType</key>
	<string>APPL</string>
	<key>CFBundleSignature</key>
	<string>????</string>
	<key>CFBundleSupportedPlatforms</key>
	<array>
		<string>MacOSX</string>
	</array>
	<key>CFBundleVersion</key>
	<string>3.7.4</string>
	<key>DTCompiler</key>
	<string>com.apple.compilers.llvm.clang.1_0</string>
	<key>DTPlatformBuild</key>
	<string>7D1014</string>
	<key>DTPlatformVersion</key>
	<string>GM</string>
	<key>DTSDKBuild</key>
	<string>15E60</string>
	<key>DTSDKName</key>
	<string>macosx10.11</string>
	<key>DTXcode</key>
	<string>0731</string>
	<key>DTXcodeBuild</key>
	<string>7D1014</string>
	<key>LSApplicationCategoryType</key>
	<string>public.app-category.games</string>
	<key>LSMinimumSystemVersion</key>
	<string>10.7</string>
	<key>NSMainNibFile</key>
	<string>MainMenu</string>
	<key>NSPrincipalClass</key>
	<string>NSApplication</string>
</dict>
</plist>
```

And the current one generated in master:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>BuildMachineOSBuild</key>
	<string>16A323</string>
	<key>CFBundleDevelopmentRegion</key>
	<string>English</string>
	<key>CFBundleIconFile</key>
	<string>FS2_Open</string>
	<key>CFBundleInfoDictionaryVersion</key>
	<string>6.0</string>
	<key>CFBundleName</key>
	<string>FS2_Open</string>
	<key>CFBundlePackageType</key>
	<string>APPL</string>
	<key>CFBundleSignature</key>
	<string>????</string>
	<key>CFBundleSupportedPlatforms</key>
	<array>
		<string>MacOSX</string>
	</array>
	<key>CFBundleVersion</key>
	<string></string>
	<key>DTCompiler</key>
	<string>com.apple.compilers.llvm.clang.1_0</string>
	<key>DTPlatformBuild</key>
	<string>8A218a</string>
	<key>DTPlatformVersion</key>
	<string>GM</string>
	<key>DTSDKBuild</key>
	<string>16A300</string>
	<key>DTSDKName</key>
	<string>macosx10.12</string>
	<key>DTXcode</key>
	<string>0800</string>
	<key>DTXcodeBuild</key>
	<string>8A218a</string>
	<key>LSApplicationCategoryType</key>
	<string>public.app-category.games</string>
	<key>LSMinimumSystemVersion</key>
	<string></string>
	<key>NSMainNibFile</key>
	<string>MainMenu</string>
	<key>NSPrincipalClass</key>
	<string>NSApplication</string>
	<key>SDL_FILESYSTEM_BASE_DIR_TYPE</key>
	<string>parent</string>
</dict>
</plist>
```

And the InfoPlist.strings from 3.7.4:

```
/* Localized versions of Info.plist keys */

CFBundleShortVersionString = "FreeSpace 2 Open version 3.7.4";
CFBundleGetInfoString = "FreeSpace 2 Open version 3.7.4, Copyright 1999 Volition, Inc. & Copyright 2005-2010 The Source Code Project.";
NSHumanReadableCopyright = "Copyright 1999 Volition, Inc. & Copyright 2005-2010 The Source Code Project.";
```

And from master:

```
/* Localized versions of Info.plist keys */

CFBundleShortVersionString = "FreeSpace 2 Open version 3.7.3";
CFBundleGetInfoString = "FreeSpace 2 Open version 3.7.3, Copyright 1999 Volition, Inc. & Copyright 2005-2010 The Source Code Project.";
NSHumanReadableCopyright = "Copyright 1999 Volition, Inc. & Copyright 2005-2010 The Source Code Project.";
```

I'm not actually sure where the 3.7.3 is even coming from.

**Comment by asarium** [comment]
The 3.7.3 is hard-coded into the file.

It looks like the version is somehow replaced when the bundle is created? I have no idea how that is handled by XCode so I'll need your help to figure out how to do this.

**Comment by chief1983** [comment]
I see, InfoPlist.strings is one of the binary files for some reason, so it is not found with 'git grep', which can't find any other places with '3.7.3' in the codebase.  I am pretty sure the regex version replacer code I had in the Perl build system could handle that particular file, but it was tricky, it's like, UTF-16 or something.  Think there's anyway the version replacement could work in that file too, or do we just need to manually update that between revisions for now?  At the least we should have this bumped to 3.7.5 by now, and keep updating it during the 3.8 process.

**Comment by chief1983** [comment]
These are the contents of my InfoPlist.strings file after compilation now:

��

I don't quite think this worked as intended.

**Comment by asarium** [comment]
The .in file is encoded using UTF-16 for some reason and even that doesn't look right when I look at it. I'll try to reencode it using UTF-8.

**Comment by chief1983** [comment]
The original file has always been utf-16, I think it's a requirement of some sort for Mac apps, not sure.  But I dealt with that with the perl replacer too, it had its own function dedicated to utf-16 replacement.

**Comment by chief1983** [comment]
So I was looking for evidence that the files have to be UTF-16, and I actually found something suggesting they don't, I think

https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/LoadingResources/Strings/Strings.html

The Note seems to indicate that we could start with a UTF-8 file and it should work just fine.  I guess we give that a shot?  There seem to be lots of people on SO recommending it should be UTF-16, but Apple itself seems to indicate otherwise, so we might as well try.

**Comment by asarium** [comment]
Have you tried the latest version of this branch? I converted the files to UTF-8 and it seems to be working (at least it compiles). The output from [Travis CI](https://travis-ci.org/scp-fs2open/fs2open.github.com/jobs/193346604#L936) also suggests that the right content is used for the strings file.

**Comment by chief1983** [comment]
Oh I missed your commit, I was doing the same thing without noticing that you had made a patch.  I was about to check my compilation result, I'll check the hash after I see if my change worked, we probably did the same thing.

**Comment by chief1983** [comment]
With my change, I end up with this:
```
Cliffs-MacBook-Pro:build cliffgordon$ file -I bin/Debug/fs2_open_3_7_5_x64-DEBUG.app/Contents/Resources/English.lproj/InfoPlist.strings 
bin/Debug/fs2_open_3_7_5_x64-DEBUG.app/Contents/Resources/English.lproj/InfoPlist.strings: text/plain; charset=us-ascii
```

But with 3.7.4, we had this:
```
Cliffs-MacBook-Pro:build cliffgordon$ file -I /Applications/Freespace2/FS2_Open\ 3.7.4.app/Contents/Resources/English.lproj/InfoPlist.strings 
/Applications/Freespace2/FS2_Open 3.7.4.app/Contents/Resources/English.lproj/InfoPlist.strings: text/plain; charset=utf-16le
```

With a master build from the other day though, we had this:
```
Cliffs-MacBook-Pro:build cliffgordon$ file -I /Applications/Freespace2/fs2_open_3_7_5_x64.app/Contents/Resources/English.lproj/InfoPlist.strings 
/Applications/Freespace2/fs2_open_3_7_5_x64.app/Contents/Resources/English.lproj/InfoPlist.strings: text/plain; charset=utf-16be
```

With your branch as it is now, I did not even get a Resource/English.lproj folder in the resulting application bundle.  That said, the only thing keeping my method from showing up as UTF-16 on the console, might be the lack of BOM.  But a BOM is not required for UTF-16 anymore, I don't think.  I know the file looked good to me anyway.  All I did was remove the existing UTF-16 .in file, create a new UTF-8 file in its place, pasted in the text that had the 3.7.3 version from a current master application bundle, and changed the 3.7.3 to the new token.  Saved that, compiled, and got what seems like a functioning result, with the correct 3.7.5 version number, and it works in Finder's Get Info menu to show the copyright and Version string.

**Comment by asarium** [comment]
I tried something different with the way how I tell CMake to copy the file to the bundle. If I revert that it should match how you did it which should fix the issue.

**Comment by asarium** [comment]
I updated the way the resource file is copied into the bundle which should be the same what your code is currently doing.

**Comment by chief1983** [comment]
It's closer, now it is putting the file as English.lproj, instead of English.lproj/InfoPlist.strings.  But the contents appear correct.  Just a level off in the location.

File: freespace2/resources.cmake [hunk 1]
```diff
@@ -51,19 +51,14 @@ elseif(APPLE)
...
-    #file(GENERATE
-    #    OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/fixup_bundle-$<CONFIG>.cmake"
-    #    INPUT "${CMAKE_CURRENT_BINARY_DIR}/fixup_bundle.cmake")
-
+    # Put the generated files into the bundle
     add_custom_command(TARGET Freespace2 POST_BUILD
 		COMMAND cp -a "${FSO_MAC_FRAMEWORKS}" "$<TARGET_FILE_DIR:Freespace2>/../Frameworks"
-        #COMMAND "${CMAKE_COMMAND}" -P "${CMAKE_CURRENT_BINARY_DIR}/fixup_bundle-$<CONFIG>.cmake"
-        COMMENT "Copying frameworks into bundle..."
+        COMMAND cp -a "${CMAKE_CURRENT_BINARY_DIR}/InfoPlist.strings" "$<TARGET_FILE_DIR:Freespace2>/../Resources/English.lproj"
```
Comment (chief1983) [inline freespace2/resources.cmake]: I think this just needs a trailing slash after the English.lproj to fix the issue now.

Comment (asarium) [inline freespace2/resources.cmake]: The `MACOSX_PACKAGE_LOCATION` property is supposed to copy the files into the bundle as part of the build but in my test that didn't work.

**Comment by chief1983** [comment]
Testing now.  FYI though, there appears to be a conflict with master that will need to be resolved.

**Comment by asarium** [comment]
I already resolved it.

**Comment by chief1983** [comment]
Ugh, my bad, I forgot to pull when I checked out your branch.  Rebuilding again so I can test the _new_ code.

**Comment by asarium** [comment]
@chief1983 Did you have a chance to test my newest changes?

**Comment by chief1983** [comment]
I thought I had and reported back before I left work Friday, but I must have gotten distracted.  I'll double check Monday.  Seem to recall seeing it having worked though.

**Comment by chief1983** [comment]
Ok, the file copying is happening correctly, but the file info dialog isn't showing the version or copyright.  After re-saving the file with Sublime as UTF-8 and not UTF-8 with BOM, it works.  I think the source file InfoPlist.strings.in just needs to have the BOM removed.  Can be done via Sublime Text, just open and Save with Encoding -> UTF-8, instead of UTF-8 with BOM.  Or create it as a UTF-8 file from scratch by copying into a new file and overwriting that one.

Sublime won't show the BOM, but when I cat the file on the terminal, there is a empty character being displayed before the opening comment /.  Saving as UTF-8 without BOM removes that hidden character, and saving again as UTF-8 with BOM puts it back, so I am assuming that is what is currently the first character in the file, and removing it seems to fix the issue I see now.

**Comment by chief1983** [comment]
Also, command line git diff makes the BOM existence really obvious.  When I run git diff origin/master, the first line of the .in file is:
+<U+FEFF><U+FEFF>/* Localized versions of Info.plist keys */

**Comment by asarium** [comment]
Alright, I think I managed to remove the BOM from the file. It wasn't even the UTF-8 BOM but it was a UTF-16 BOM which may explain the issues you were encountering.

**Pull request merged**